torchrl/envs/batched_envs.py

```diff
@@ -635,14 +638,14 @@ def _reset(self, tensordict: TensorDictBase, **kwargs) -> TensorDictBase:
             self.shared_tensordicts[i].update_(
                 _td.select(*self._selected_reset_keys, strict=False)
             )
-
         if self._single_task:
             # select + clone creates 2 tds, but we can create one only
             out = TensorDict(
                 {}, batch_size=self.shared_tensordict_parent.shape, device=self.device
             )
             for key in self._selected_reset_keys:
-                if _unravel_key_to_tuple(key)[-1] != "_reset":
+                key = unravel_keys(key)
+                if key not in self.reset_keys:
```

Comment:
the reset key should not be removed here

Reply:
why? Can you explain? It used to be removed (see diff) and i'm pretty sure tests fail if you don't.

Reply:
there are many reasons we discussed

- they should alwayes be removed by `env.reset()`
- this removes them only when `self._single_task` and not otherwise

Reply:
``> If the point is that it should also be excluded inf the case where there is more than 1 task I totally agree!

definitely let's uniform this.

my other point is just that i thought we decided that the convention was that all the _resets are removed only in `env.reset()` and in this particular instance they are removed bofore. this has 2 problems:

- `env.reset()` cannot perform the done check appropriately (it will see a missing _reset and assume it is all true but that is not the case)(this is a problem if the reset was partial and some dones are true after the reset as they should but they cause the check to fail)
-  and that this is not uniform as the batched env _reset() will be the only one to not provide _resets

Reply:
> env.reset() cannot perform the done check appropriately (it will see a missing _reset and assume it is all true but that is not the case)(this is a problem if the reset was partial and some dones are true after the reset as they should but they cause the check to fail)

no because env.reset gets its "_reset" from the input tensordict, not the output

> and that this is not uniform as the batched env _reset() will be the only one to not provide _resets

Same I don't think it is true: per se all _reset must return a new tensordict and it's written nowhere in the contract of that function that you must copy any part of the content of the input in the output. In other words: we never read "_reset" from the output of _reset. Check the `reset` method for more context if need reassurance.

Reply:
ok got it, my bad.

let's just test that BatchedEnv._reset() does not return _resets in all branches